core/network-libp2p/src/custom_proto/behaviour.rs

```diff
@@ -372,10 +372,33 @@ impl<TMessage, TSubstream> CustomProtos<TMessage, TSubstream> {
 				.count();
 			self.max_outgoing_connections - num_outgoing_connections
 		};
-
 		trace!(target: "sub-libp2p", "Connect-to-nodes round; attempting to fill {:?} slots",
 			num_to_open);
 
+		// We first try to enable existing connections.
+		for peer_id in &self.connected_peers {
+			if num_to_open == 0 {
+				break
+			}
+
+			if self.enabled_peers.contains_key(peer_id) {
+				continue;
+			}
+
+			if let Some((_, expire)) = self.banned_peers.iter().find(|(p, _)| p == peer_id) {
+				if *expire >= Instant::now() {
+					continue;
+				}
+			}
+
+			trace!(target: "sub-libp2p", "Enabling custom protocols with {:?} (active)", peer_id);
+			self.events.push(NetworkBehaviourAction::SendEvent {
```

Comment:
If we enabled this "old" connection, do we still need to try to open a new connection?

Reply:
That doesn't open a new connection. It sends a message to the handler of the existing connection that it should enable it.

Reply:
What I really wanted to ask, shouldn't we return after sending this event?

Reply:
We continue looping until `num_to_open` is 0.